internal/indexer/processors/contract_deploy.go

```diff
@@ -0,0 +1,110 @@
+package processors
+
+import (
+	"context"
+	"fmt"
+
+	operation_processor "github.com/stellar/go/processors/operation"
+	"github.com/stellar/go/xdr"
+
+	"github.com/stellar/wallet-backend/internal/indexer/types"
+)
+
+// ContractDeployProcessor emits state changes for contract deployments.
+type ContractDeployProcessor struct {
+	networkPassphrase string
+}
+
+func NewContractDeployProcessor(networkPassphrase string) *ContractDeployProcessor {
+	return &ContractDeployProcessor{networkPassphrase: networkPassphrase}
+}
+
+// ProcessOperation emits a state change for each contract deployment (including subinvocations).
+func (p *ContractDeployProcessor) ProcessOperation(_ context.Context, op *operation_processor.TransactionOperationWrapper) ([]types.StateChange, error) {
+	if op.OperationType() != xdr.OperationTypeInvokeHostFunction {
+		return nil, ErrInvalidOpType
+	}
+	invokeHostOp := op.Operation.Body.MustInvokeHostFunctionOp()
+
+	opID := op.ID()
+	builder := NewStateChangeBuilder(op.Transaction.Ledger.LedgerSequence(), op.LedgerClosed.Unix(), op.Transaction.Hash.HexString()).
+		WithOperationID(opID).
+		WithCategory(types.StateChangeCategoryContract).
+		WithReason(types.StateChangeReasonDeploy)
+
+	deployedContractsMap := map[string]types.StateChange{}
+
+	processCreate := func(fromAddr xdr.ContractIdPreimageFromAddress) error {
+		contractID, err := calculateContractID(p.networkPassphrase, fromAddr)
+		if err != nil {
+			return fmt.Errorf("calculating contract ID: %w", err)
+		}
+		deployerAddr, err := fromAddr.Address.String()
+		if err != nil {
+			return fmt.Errorf("deployer address to string: %w", err)
+		}
+
+		deployedContractsMap[contractID] = builder.Clone().WithAccount(contractID).WithDeployer(deployerAddr).Build()
+		return nil
+	}
+
+	var walkInvocation func(inv xdr.SorobanAuthorizedInvocation) error
+	walkInvocation = func(inv xdr.SorobanAuthorizedInvocation) error {
```

Comment:
I didn't realize this `walkInvocation` pattern was [provided by the JS SDK](https://stellar.github.io/js-stellar-sdk/global.html#walkInvocationTree). I wonder if we can request this to be added to the txnbuild golang package?

Reply:
`walkInvocation` is a common pattern to traverse node-like structures like sub-invocations.